### PR TITLE
adding support to KBallSearcher

### PR DESCRIPTION
**Pull request by exepulveda**
Implementing KBallSearcher as we discussed. 
KBallSearcher uses knn and radius combined by searching for knn first and then filtering the resutls to locations withtin the radius.

### CODE REVIEW

**Comment by juliohm** [comment]
Depends on https://github.com/JuliaEarth/GeoStatsBase.jl/pull/44

**Comment by juliohm** [comment]
@exepulveda thanks for tackling this one, I saw you spotted some interesting optimizations with Symmetric in this PR. Should we open two separate PRs, one for the new KBallSearcher improvements and another for the Symmetric optimizations?

Please let me know if we can close this one, and start new PRs from the original JuliaEarth/KrigingEstimators.jl repo instead of forks now that you can create branches in the org. Looking forward to improving the code further.

**Comment by exepulveda** [comment]
Yes! We can close it now. I did the test and them passed. I also opened a new issue to address the symmetric properties of kriging linear systems.

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl]: Perhaps we can erase the comment, it is inferred from the code directly. Also, my pedantic mode asks for an extra space after the last comma, could you please fix these minor details?

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl]: Extra space after last comma is missing.

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl:124]: Should we keep the path here for consistency? I noticed you moved it to outside the if-else branch?

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl]: It is so good to see this code going away! :100:

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl:188]: Just a comment that in a future release we should perhaps consider only `missing` as the sentinel. The `NaN` here is a temporary hack I've adopted, but certainly deserves a future review :+1:

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl:194]: Nice!

**Review by juliohm — CHANGES_REQUESTED** [review]
That is a great PR @exepulveda. Could you please make some additional minor changes before we merge it? Looking forward to the next round of reviews.

File: src/solvers/kriging.jl
Comment (exepulveda) [reply src/solvers/kriging.jl]: There are many comma missing around the code. Should I take care of the code that I just wrote?

File: src/solvers/kriging.jl
Comment (exepulveda) [reply src/solvers/kriging.jl:124]: I believe we just need LinearPath before creating preproc. Although it seems path is not longer needed, for not sure. Maybe this could be done later.

Comment (juliohm) [reply src/solvers/kriging.jl:124]: Nice, path is also gone :100: it was needed before because of the `SourcePath` trick. Glad we don't need it anymore here.

File: src/KrigingEstimators.jl
Comment (juliohm) [inline src/KrigingEstimators.jl]: Could you please revert this change? I understood we will be including these optimizations in a separate PR right?

File: src/solvers/kriging.jl
Comment (juliohm) [inline src/solvers/kriging.jl]: Given that we always use a linear path, maybe we can erase it from the `preprocess` function entirely.

**Review by juliohm — CHANGES_REQUESTED** [review]
Nice, thank you for improving the PR. I've added a few minor changes that I think will make it ready to merge into master :+1: Basically we can remove the `path` from the `preproc` dict now and stop using it in the `solve` function. We can instantiate the `LinearPath` explicitly in the `solve` function.

File: src/solvers/kriging.jl
Comment (juliohm) [reply src/solvers/kriging.jl]: Please feel free to fix any missing comma you find in the source.

**Comment by exepulveda** [comment]
HI @juliohm,
My apologies. I pushed to the master branch instead of to the exequiel_dev branch. May bad. To be honest, first time working an a big project and git. Having such rights is dangerous :-) 

I reverted the commit to do it properly by PR from the exequiel_dev. I hope not to have caused major trouble.

The proposal of Symmetry in OK is in the exequiel_dev branch waiting for you approval of reverted commit.

**Comment by juliohm** [comment]
No worries, it happens sometimes, we just need to be careful to remember :+1: 

> Having such rights is dangerous :-)

I wouldn't give such rights to anyone, I know you have good programming skills :)

**Comment by juliohm** [comment]
I looked into the `exequiel_dev` branch and there are no changes compared to master, could you please double check?

**Comment by exepulveda** [comment]
That is strange. I have checked both versions of KrigingEstimators.jl
Line 10 in master:
`using LinearAlgebra: Factorization, lu, cholesky, issuccess, ⋅`
Line 10 in exequiel_dev: 
`using LinearAlgebra: Factorization, Symmetric, bunchkaufman, lu, cholesky, issuccess, ⋅`

They are different.

**Comment by juliohm** [comment]
It is probably a delay in github then. I will try to refresh the page to
see if the changes appear again.

**Comment by juliohm** [comment]
I think it is because the branch is behind master (due to the revert in master):

image

Could you please start the work in a separate branch? Maybe it is easy to just delete `exequiel_dev` and start afresh to avoid git-related issues.

File: src/solvers/kriging.jl
Comment (juliohm) [reply src/solvers/kriging.jl]: @exepulveda could you please check this one? I think we can just get rid of path if we are always using a linear path. In estimation loop we can just write `traverse(pdomain, LinearPath())`

**Review by juliohm — COMMENTED** [review]
It would be nice if you could explain the new behavior. I noticed that the NearestKriging1D.png test image has changed, but I couldn't understand the plotted curve. Also, could you please add some tests for the new dataset you added to the test suite (data3d.csv)?

**Comment by exepulveda** [comment]
I would need more time to do more tests for 3D cases. I think we need to merge with master the changes from the fork exepulveda:kballsearcher. Once this is done, I will delete this fork and work with developing branches. Is that make sense?

regarding the image. Not clear what I did for this image to change. The tests produce several images that are hard to understand in terms of testing. Maybe accidentally I replaced that one in the many tests.

**Comment by juliohm** [comment]
> I think we need to merge with master the changes from the fork exepulveda:kballsearcher. Once this is done, I will delete this fork and work with developing branches. Is that make sense?

As soon as we have a PR with the changes you are discussing, it makes sense to me. I can't see another PR besides this one we are currently reviewing.

**Comment by juliohm** [comment]
There are some pending review comments above. We can't merge without carefully checking the change in the test image above. If you have any questions regarding the images in the test suite, please share here and we can double check.

Also, the 3D dataset, if you don't have a plan to add tests for it in this PR, then it shouldn't be included here.

**Comment by exepulveda** [comment]
Let me recap the pendings:

1. Image for test has changed
2. data3d.csv should not be commited until the tests are also commited (I will deleted it)
3. Yo want to move LinearPath()

Please confirm.

**Comment by juliohm** [comment]
On the spot 👍

**Comment by juliohm** [comment]
@exepulveda I've fixed the issue with `valid` on the master branch, you can rebase this PR on top of the latest master :+1:

**Comment by exepulveda** [comment]
Thanks. I hope to finish this PR today or so. My developing environment is at uni :-)

**Comment by exepulveda** [comment]
I hope this time we can merge to master. Waiting for you to continue with other things.

**Comment by juliohm** [comment]
Awesome :100: , I am merging it now assuming that all tests passed locally for you. The tests are failing in Travis CI because it was using a version of Julia (Julia v1.3) we dropped support recently on the master branch.

Thank you again for this nice improvement. :)

**Comment by exepulveda** [comment]
Indeed. All tests passed locally. 
Also, I did the tests with v1.5 as well (just for KrigingEstimators.jl).

**Comment by exepulveda** [comment]
There is still something the double check. With the filtering of missing, the kdtree should also be built only with valid samples and the search accordingly.

**Comment by juliohm** [comment]
I've recently updated `.travis.yml` to test on Julia v1.4, v1.5 and the nightly build. I've also added [this workaround](https://github.com/JuliaEarth/KrigingEstimators.jl/blob/3e5933947376c0b7c0e667821c269b2c614b9ad8/src/solvers/kriging.jl#L231-L235) to filter the missing values while https://github.com/JuliaEarth/GeoStatsBase.jl/pull/48 is not merged. Appreciate if you can review the PR. I can merge it right away if everything is ok, and then update the Kriging solver again to discard missings everywhere.

**Comment by exepulveda** [comment]
Yeap. I saw it. It is ok. Let me know when is merged and I can continue with Kriging solver after.

**Comment by juliohm** [comment]
Everything is merged. Feel free to continue the work 👍

**Pull request merged**
